File: Nov2020/20Nov/BaseballGame.cpp
/*
Problem name: Baseball Game
Problem link: https://leetcode.com/problems/baseball-game/
*/

class Solution {
public:
    int calPoints(vector<string>& ops) {
        int sum=0,i=0,t1=0,t2=0;
        stack<int> record;
        
        while(i<ops.size()){
            if(ops[i]=="+"){
                t1=record.top();
                record.pop();
                t2=record.top();
                record.push(t1);                
                record.push(t1+t2);
                sum+=t1+t2;
            }
            else if(ops[i]=="D"){
                sum+=2*record.top();
                record.push(2*record.top());
                
            }
            else if(ops[i]=="C"){
                sum-=record.top();
                record.pop();
            }
            else{
                record.push(stoi(ops[i]));
                sum+=stoi(ops[i]);
            }
            i++;
        }
        return sum;
        
    }
};

File: Nov2020/20Nov/BaseballGame.java
/*
Problem Name: 682. Baseball Game
Problem Link: https://leetcode.com/problems/baseball-game/
*/
class Solution {
    public int calPoints(String[] ops) {
        Stack<Integer> st = new Stack<Integer>();
        for(String s: ops){
            if(s.equals("C"))
                st.pop();
            else if(s.equals("D"))
                st.push(st.peek()*2);
            else if(s.equals("+")){
                int t1 = st.pop(), t2 = st.pop();
                st.push(t2);
                st.push(t1);
                st.push(t1 + t2);
            }
            else
                st.push(Integer.parseInt(s));
        }
        int sum = 0;
        while(!st.isEmpty())
            sum += st.pop();
        return sum;
    }
}
